core/src/main/java/io/grpc/inprocess/InProcessTransport.java

```diff
@@ -115,20 +115,32 @@ public void run() {
   }
 
   @Override
-  public synchronized ClientStream newStream(MethodDescriptor<?, ?> method,
-      Metadata headers, ClientStreamListener clientStreamListener) {
+  public synchronized ClientStream newStream(
+      final MethodDescriptor<?, ?> method, final Metadata headers) {
     if (shutdownStatus != null) {
-      clientStreamListener.closed(shutdownStatus, new Metadata());
-      return new NoopClientStream();
+      return new NoopClientStream() {
+        @Override
+        public void start(ClientStreamListener listener) {
+          listener.closed(shutdownStatus, new Metadata());
+        }
+      };
     }
-    headers.removeAll(GrpcUtil.AUTHORITY_KEY);
-    InProcessStream stream = new InProcessStream();
-    stream.serverStream.setListener(clientStreamListener);
-    ServerStreamListener serverStreamListener = serverTransportListener.streamCreated(
-        stream.serverStream, method.getFullMethodName(), headers);
-    stream.clientStream.setListener(serverStreamListener);
-    streams.add(stream);
-    return stream.clientStream;
+
+    class StartCallback implements Runnable {
```

Comment:
Since this callback is the only possible callback passed to `InProcessStream`, I would prefer avoiding the callback mechanism and moving the work into `InPorcessStream.start()`. It would require passing `method` and `headers` to `InProcessStream`, but it's better than `StartCallback` and `InProcessStream` being tangled with each other.